kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,994 @@
+                            };
+                            let val_name = if let Some(val_name) = key_val_names.next() {
+                                val_name
+                            } else {
+                                return Err(Error::generic("map fields didn't include a val col"));
+                            };
+                            if key_val_names.next().is_some() {
+                                return Err(Error::generic("map fields had more than 2 members"));
+                            }
+                            let inner_schema = map_type.as_struct_schema(key_name, val_name);
+                            let (parquet_advance, _children) = get_indices(
+                                parquet_index + parquet_offset,
+                                &inner_schema,
+                                inner_fields,
+                                mask_indices,
+                            )?;
+                            // advance the number of parquet fields, but subtract 1 because the
+                            // map will be counted by the `enumerate` call but doesn't count as
+                            // an actual index.
+                            parquet_offset += parquet_advance - 1;
+                            // note that we found this field
+                            found_fields.insert(requested_field.name());
+                            // push the child reorder on, currently no reordering for maps
+                            reorder_indices.push(ReorderIndex::Index { index });
+                        }
+                        _ => {
+                            return Err(Error::unexpected_column_type(field.name()));
+                        }
+                    }
+                }
+            }
+            _ => {
+                if let Some((index, _, requested_field)) =
+                    requested_schema.fields.get_full(field.name())
+                {
+                    found_fields.insert(requested_field.name());
+                    mask_indices.push(parquet_offset + parquet_index);
+                    reorder_indices.push(ReorderIndex::Index { index });
+                }
+            }
+        }
+    }
+    if found_fields.len() != requested_schema.fields.len() {
+        // some fields are missing, but they might be nullable, need to insert them into the reorder_indices
+        for (requested_position, field) in requested_schema.fields().enumerate() {
+            if !found_fields.contains(field.name()) {
+                if field.nullable {
+                    debug!("Inserting missing and nullable field: {}", field.name());
+                    reorder_indices.push(ReorderIndex::Null {
+                        index: requested_position,
+                        field: Arc::new(field.try_into()?),
+                    });
+                } else {
+                    return Err(Error::Generic(format!(
```

Comment:
I did rename `Null` to missing. But I think in this context we do want to return an error, because you're asking us to read something that's not there. If it's something with a default value, it shouldn't be passed in the requested schema.

Perhaps at some point we could support passing default values in here, and then this code could fill it in, but forcing callers to deal with `Missing` at any arbitrary depth in a complex parquet record batch seems... not so nice.

I'd rather just only have this code support adding in the null columns for nullable stuff, and punt defaults to the future, if they are needed.